README.md
```
'   :::===  :::  ===  === ::: :::==== :::===== :::  ===
'   :::     :::  ===  === ::: :::==== :::      :::  ===
'    =====  ===  ===  === ===   ===   ===      ========
'       ===  ===========  ===   ===   ===      ===  ===
'   ======    ==== ====   ===   ===    ======= ===  ===
'                                                              
```

# Switch

[![CI Status](https://img.shields.io/travis/T-Pham/Switch/master.svg?style=flat-square)](https://travis-ci.org/T-Pham/Switch)
[![GitHub issues](https://img.shields.io/github/issues/T-Pham/Switch.svg?style=flat-square)](https://github.com/T-Pham/Switch/issues)
[![Codecov](https://img.shields.io/codecov/c/github/T-Pham/Switch.svg?style=flat-square)](https://codecov.io/gh/T-Pham/Switch)
[![Documentation](https://img.shields.io/cocoapods/metrics/doc-percent/RoundedSwitch.svg?style=flat-square)](http://cocoadocs.org/docsets/RoundedSwitch)

[![GitHub release](https://img.shields.io/github/tag/T-Pham/Switch.svg?style=flat-square&label=release)](https://github.com/T-Pham/Switch/releases)
[![Platform](https://img.shields.io/cocoapods/p/RoundedSwitch.svg?style=flat-square)](https://github.com/T-Pham/Switch)
[![License](https://img.shields.io/cocoapods/l/Switch.svg?style=flat-square)](LICENSE)

[![SwiftPM](https://img.shields.io/badge/SwiftPM-compatible-4BC51D.svg?style=flat-square)](https://swift.org/package-manager)
[![Carthage](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat-square)](https://github.com/Carthage/Carthage)
[![CocoaPods](https://img.shields.io/badge/CocoaPods-compatible-4BC51D.svg?style=flat-square)](https://cocoapods.org/pods/RoundedSwitch)

## Description

An iOS switch control implemented in Swift with full Interface Builder support.

To run the demo project:

`pod try RoundedSwitch`

![Switch](https://github.com/T-Pham/Switch/blob/master/switch.gif?raw=true)

![Switch](https://github.com/T-Pham/Switch/blob/master/switch.png?raw=true)

## Usage

Either config the switch in the Interface Builder or programatically as follow:

```swift
import Switch
...
let mySwitch = Switch()
mySwitch.leftText = "Windows"
mySwitch.rightText = "Mac"
mySwitch.rightSelected = true
mySwitch.tintColor = UIColor.purple
mySwitch.disabledColor = mySwitch.tintColor.withAlphaComponent(0.4)
mySwitch.backColor = mySwitch.tintColor.withAlphaComponent(0.05)
mySwitch.sizeToFit()
mySwitch.addTarget(self, action: #selector(ViewController.switchDidChangeValue(_:)), for: .valueChanged)
```

Please note that the module name is `Switch`. However, when installed with CocoaPods, it is `RoundedSwitch`.

Please see the [Reference Documentation](http://cocoadocs.org/docsets/RoundedSwitch) for details.

## Installation

### [Swift Package Manager](https://swift.org/package-manager)

In Xcode, select `File->Swift Packages->Add Package Dependency...` and follow the instructions on-screen.
Use the following repo URL when prompted.

```
https://github.com/T-Pham/Switch
```

### [Carthage](https://github.com/Carthage/Carthage)

Add the line below to your Cartfile:

```ruby
github "T-Pham/Switch"
```

### [CocoaPods](https://cocoapods.org/pods/RoundedSwitch)

Add the line below to your Podfile:

```ruby
pod 'RoundedSwitch'
```

### Manually

Add the file [`Switch.swift`](Switch.swift) to your project. You are all set.

## Compatibility
From version 2.0.0, Swift 3 syntax is used. If your project is still using Swift version 2, please use a UITextField-Navigation version prior to 2.0.0.

Podfile

```ruby
pod 'RoundedSwitch', '~> 1.0.3'
```

or Cartfile

```ruby
github "T-Pham/Switch" ~> 1.0.3
```

## License

Switch is available under the MIT license. See the [LICENSE](LICENSE) file for more info.


CHANGELOG.md
# v1.0.0
Initial release

# v1.0.1
Fix - pod try does not work

# v1.0.2
Add documentation

# v1.0.3
Add some tests

# v2.0.0
Upgrade to Swift 3

# v2.0.1
Add backColor property

# v2.0.2
Fix - very hard to switch by tapping

# v3.0.0
Update to Swift 5


Sources/Switch/Switch.swift
//
//  Switch.swift
//
//  Created by Thanh Pham on 8/24/16.
//

import UIKit

/// A switch control
@IBDesignable open class Switch: UIControl {

    let backgroundLayer = RoundLayer()
    let switchLayer = RoundLayer()

    var previousPoint: CGPoint?
    var switchLayerLeftPosition: CGPoint?
    var switchLayerRightPosition: CGPoint?

    static let labelFactory: () -> UILabel = {
        let label = UILabel()
        label.textAlignment = .center
        return label
    }

    /// The label on the left. Don't set the `textColor` and `text` properties on this label. Set them on the `leftText`, `tintColor` and `disabledColor` properties of the switch instead.
    public let leftLabel = labelFactory()

    /// The label on the right. Don't set the `textColor` and `text` properties on this label. Set them on the `rightText`, `tintColor` and `disabledColor` properties of the switch instead.
    public let rightLabel = labelFactory()

    /// Text for the label on the left. Setting this property instead of the `text` property of the `leftLabel` to trigger Auto Layout automatically.
    @IBInspectable open var leftText: String? {
        get {
            return leftLabel.text
        }
        set {
            leftLabel.text = newValue
            invalidateIntrinsicContentSize()
        }
    }

    /// Text for the label on the right. Setting this property instead of the `text` property of the `rightLabel` to trigger Auto Layout automatically.
    @IBInspectable open var rightText: String? {
        get {
            return rightLabel.text
        }
        set {
            rightLabel.text = newValue
            invalidateIntrinsicContentSize()
        }
    }

    /// True when the right value is selected, false when the left value is selected. When this property is changed, the UIControlEvents.ValueChanged is fired.
    @IBInspectable open var rightSelected: Bool = false {
        didSet {
            reloadSwitchLayerPosition()
            reloadLabelsTextColor()
            sendActions(for: .valueChanged)
        }
    }

    /// The color used for the unselected text and the border. The default value is light gray.
    @IBInspectable open var disabledColor: UIColor = UIColor.lightGray {
        didSet {
            backgroundLayer.borderColor = disabledColor.cgColor
            reloadLabelsTextColor()
        }
    }

    /// The color used for the switch background. Transparent by default.
    @IBInspectable open var backColor: UIColor = UIColor.clear {
        didSet {
            backgroundLayer.backgroundColor = backColor.cgColor
        }
    }

    /// The color used for the selected text and the switch border.
    override open var tintColor: UIColor! {
        didSet {
            switchLayer.borderColor = tintColor.cgColor
            reloadLabelsTextColor()
        }
    }

    var touchBegan = false
    var touchBeganInSwitchLayer = false
    var touchMoved = false

    /// Init with coder.
    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setup()
    }

    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }

    func setup() {
        backgroundLayer.backgroundColor = UIColor.white.cgColor
        backgroundLayer.borderColor = disabledColor.cgColor
        backgroundLayer.borderWidth = 1
        layer.addSublayer(backgroundLayer)

        switchLayer.borderColor = tintColor.cgColor
        switchLayer.borderWidth = 1
        layer.addSublayer(switchLayer)

        addSubview(leftLabel)
        addSubview(rightLabel)
        reloadLabelsTextColor()
    }

    /// Layouts subviews. Should not be called directly.
    override open func layoutSubviews() {
        super.layoutSubviews()
        leftLabel.frame = CGRect(x: 0, y: 0, width: bounds.size.width / 2, height: bounds.size.height)
        rightLabel.frame = CGRect(x: bounds.size.width / 2, y: 0, width: bounds.size.width / 2, height: bounds.size.height)
    }

    /// Layouts sublayers of a layer. Should not be called directly.
    override open func layoutSublayers(of layer: CALayer) {
        super.layoutSublayers(of: layer)
        guard layer == self.layer else {
            return
        }
        backgroundLayer.frame = layer.bounds
        switchLayer.bounds = CGRect(x: 0, y: 0, width: layer.bounds.size.width / 2, height: layer.bounds.size.height)
        switchLayerLeftPosition = layer.convert(layer.position, from: layer.superlayer).applying(CGAffineTransform(translationX: -switchLayer.bounds.size.width / 2, y: 0))
        switchLayerRightPosition = switchLayerLeftPosition!.applying(CGAffineTransform(translationX: switchLayer.bounds.size.width, y: 0))
        touchBegan = false
        touchBeganInSwitchLayer = false
        touchMoved = false
        previousPoint = nil
        reloadSwitchLayerPosition()
    }

    /// Touches began event handler. Should not be called directly.
    override open func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesBegan(touches, with: event)
        touchBegan = true
        previousPoint = touches.first!.location(in: self)
        touchBeganInSwitchLayer = switchLayer.contains(switchLayer.convert(previousPoint!, from: layer))
    }

    /// Touches moved event handler. Should not be called directly.
    override open func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesMoved(touches, with: event)
        touchMoved = true
        guard touchBeganInSwitchLayer else {
            return
        }
        let point = touches.first!.location(in: self)
        var position = switchLayer.position.applying(CGAffineTransform(translationX: point.x - previousPoint!.x, y: 0))
        position.x = max(switchLayer.bounds.size.width / 2, min(layer.bounds.size.width - switchLayer.bounds.size.width / 2, position.x))
        switchLayer.position = position
        previousPoint = point
    }

    /// Touches ended event handler. Should not be called directly.
    override open func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesEnded(touches, with: event)
        guard touchBegan else {
            return
        }
        previousPoint = nil
        if touchBeganInSwitchLayer {
            if touchMoved {
                rightSelected = abs(switchLayerLeftPosition!.x - switchLayer.position.x) > abs(switchLayerRightPosition!.x - switchLayer.position.x)
            }
        } else {
            let point = touches.first!.location(in: self)
            let touchEndedInLayer = layer.contains(point)
            let touchEndedInSwitchLayer = switchLayer.contains(switchLayer.convert(point, from: layer))
            let touchEndedInInactiveSide = touchEndedInLayer && !touchEndedInSwitchLayer
            if touchEndedInInactiveSide {
                rightSelected = !rightSelected
            }
        }
        touchBegan = false
        touchBeganInSwitchLayer = false
        touchMoved = false
    }

    /// Touches cancelled event handler. Should not be called directly.
    override open func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesCancelled(touches, with: event)
        previousPoint = nil
        reloadSwitchLayerPosition()
        touchBegan = false
        touchBeganInSwitchLayer = false
        touchMoved = false
    }

    /// Returns the minimum size that fits a given size.
    override open func sizeThatFits(_ size: CGSize) -> CGSize {
        let labelSize = CGSize(width: (size.width - 3 * size.height / 2) / 2, height: size.height)
        return desiredSizeForLeftSize(leftLabel.sizeThatFits(labelSize), rightSize: rightLabel.sizeThatFits(labelSize))
    }

    /// The minimum size used for Auto Layout.
    override open var intrinsicContentSize : CGSize {
        return desiredSizeForLeftSize(leftLabel.intrinsicContentSize, rightSize: rightLabel.intrinsicContentSize)
    }

    func desiredSizeForLeftSize(_ leftSize: CGSize, rightSize: CGSize) -> CGSize {
        let height = max(leftSize.height, rightSize.height)
        return CGSize(width: max(leftSize.width, rightSize.width) * 2 + 3 * height / 2, height: height)
    }

    func reloadLabelsTextColor() {
        leftLabel.textColor = rightSelected ? disabledColor : tintColor
        rightLabel.textColor = rightSelected ? tintColor : disabledColor
    }

    func reloadSwitchLayerPosition() {
        guard let switchLayerLeftPosition = switchLayerLeftPosition, let switchLayerRightPosition = switchLayerRightPosition else {
            return
        }
        switchLayer.position = rightSelected ? switchLayerRightPosition : switchLayerLeftPosition
    }
}

class RoundLayer: CALayer {

    override func layoutSublayers() {
        super.layoutSublayers()
        cornerRadius = bounds.size.height / 2
    }
}


SwitchTests/SwitchTests.swift
//
//  SwitchTests.swift
//  SwitchTests
//
//  Created by Thanh Pham on 9/9/16.
//  Copyright © 2016 TPM. All rights reserved.
//

import XCTest
@testable import Switch

class SwitchTests: XCTestCase {

    var theSwitch: Switch!

    override func setUp() {
        super.setUp()
        theSwitch = Switch()
    }

    override func tearDown() {
        theSwitch = nil
        super.tearDown()
    }

    func testLeftText() {
        let text = "left"
        theSwitch.leftText = text
        XCTAssertEqual(theSwitch.leftLabel.text, text)
    }

    func testRightText() {
        let text = "right"
        theSwitch.leftText = text
        XCTAssertEqual(theSwitch.leftLabel.text, text)
    }

    func testDisabledColor() {
        let color = UIColor.cyan
        theSwitch.disabledColor = color
        XCTAssertEqual(theSwitch.backgroundLayer.borderColor, color.cgColor)
        XCTAssertEqual(theSwitch.rightLabel.textColor, color)
    }

    func testTintColor() {
        let color = UIColor.purple
        theSwitch.tintColor = color
        XCTAssertEqual(theSwitch.switchLayer.borderColor, color.cgColor)
        XCTAssertEqual(theSwitch.leftLabel.textColor, color)
    }

    func testBackColor() {
        let color = UIColor.brown
        theSwitch.backColor = color
        XCTAssertEqual(theSwitch.backgroundLayer.backgroundColor, color.cgColor)
    }
}


Demo/SwitchDemo/AppDelegate.swift
//
//  AppDelegate.swift
//  SwitchDemo
//
//  Created by Thanh Pham on 8/31/16.
//  Copyright © 2016 TPM. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



Demo/SwitchDemo/ViewController.swift
//
//  ViewController.swift
//  SwitchDemo
//
//  Created by Thanh Pham on 8/31/16.
//  Copyright © 2016 TPM. All rights reserved.
//

import UIKit
import Switch

class ViewController: UIViewController {

    @IBOutlet var statusLabel: UILabel!
    @IBOutlet var sectionHeaderLabel: UILabel!
    var mySwitch: Switch!

    override func viewDidLoad() {
        super.viewDidLoad()
        mySwitch = Switch()
        mySwitch.leftText = "Windows"
        mySwitch.rightText = "Mac"
        mySwitch.rightSelected = true
        mySwitch.tintColor = UIColor.purple
        mySwitch.disabledColor = mySwitch.tintColor.withAlphaComponent(0.4)
        mySwitch.backColor = mySwitch.tintColor.withAlphaComponent(0.05)
        mySwitch.sizeToFit()
        mySwitch.addTarget(self, action: #selector(ViewController.switchDidChangeValue(_:)), for: .valueChanged)
        view.addSubview(mySwitch)
    }

    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        mySwitch.frame = (sectionHeaderLabel.frame).applying(CGAffineTransform(translationX: 0, y: sectionHeaderLabel.frame.size.height + 15))
    }

    @IBAction func switchDidChangeValue(_ theSwitch: Switch) {
        statusLabel.text = "\"\(theSwitch.rightSelected ? theSwitch.rightText ?? "" : theSwitch.leftText ?? "")\" selected"
    }
}
